pallets/migrations/src/lib.rs

```diff
@@ -103,6 +115,88 @@ pub mod pallet {
 
 			weight
 		}
+
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<(), &'static str> {
+			use frame_support::traits::OnRuntimeUpgradeHelpersExt;
+
+			let mut failed = false;
+			for migration in &T::MigrationsList::get() {
+				let migration_name = migration.friendly_name();
+				let migration_name_as_bytes = migration_name.as_bytes();
+
+				let migration_done = <MigrationState<T>>::get(migration_name_as_bytes);
+				if migration_done {
+					continue;
+				}
+				log::trace!("invoking pre_upgrade() on migration {}", migration_name);
+
+				// dump the migration name to temp storage so post_upgrade will know which
+				// migrations were performed (as opposed to skipped)
```

Comment:
In what situation would a migration be skipped?

Reply:
After multiple runtime upgrades.

Say we have migrations `a, b, c` that go into runtime release `1`. 
Runtime release `2` comes along with migration `d`.

In this case, we would skip `a, b, c` because they were done in the previous release and we would perform `d` because it's new. Our fns would be invoked in this order:

`pre_upgrade()` -- here we know we need `d` but not `a, b, c`
`migrate()` -- same here, but now we're going to mark `d` as "done"
`post_upgrade()` -- by this point, all of `a, b, c, d` appear done, we don't know which were performed

If you feel like that explanation is worth adding to the comments, I'll do so. I wanted to be clear about this in my code comments but also not obnoxiously verbose :)

Reply:
Oh, duh, how did I forget that. I think that is explained well in your original PR description already.